petsc@3.19.1 +rocm w/ ROCm 5.3.0: rocprim/device/detail/device_merge.hpp:133:9: error: no matching function for call to 'merge_path'

### Steps to reproduce the issue

`petsc@3.19.1 +rocm` build is failing for me on `OLCF Frontier` using:
* `spack@develop` (a94d18ad087038f10c9a82ebbb1a1a1dddda51e8 from `Wed May 3 05:26:50 2023 -0700`)
* PrgEnv-gnu, GCC 11.2.0
* ROCm 5.3.0 (externally provided via `rocm/5.3.0` module, definitions included in Spack environment below)

Spack environment: [spack.yaml.txt](https://github.com/spack/spack/files/11391031/spack.yaml.txt)

<details><summary>Concretization</summary>
<pre>
 -   3xruui4  petsc@3.19.1%gcc@11.2.0~X+batch~cgns~complex~cuda~debug+double~exodusii~fftw+fortran~giflib+hdf5~hpddm~hwloc+hypre~int64~jpeg~knl~kokkos~libpng~libyaml~memkind+metis~mkl-pardiso~mmg~moab~mpfr+mpi~mumps~openmp~p4est~parmmg~ptscotch~random123+rocm~saws~sca
lapack+shared~strumpack~suite-sparse+superlu-dist~tetgen~trilinos~valgrind amdgpu_target=gfx90a build_system=generic clanguage=C arch=linux-sles15-zen3
[+]  vw5wbnb      ^cray-mpich@8.1.23%gcc@11.2.0+wrappers build_system=generic arch=linux-sles15-zen3
[+]  yb3vc2g      ^diffutils@3.9%gcc@11.2.0 build_system=autotools arch=linux-sles15-zen3
[+]  hr2sqld          ^libiconv@1.17%gcc@11.2.0 build_system=autotools libs=shared,static arch=linux-sles15-zen3
[+]  rx7nx3z      ^hdf5@1.14.0%gcc@11.2.0~cxx+fortran+hl~ipo~java~map+mpi+shared~szip~threadsafe+tools api=default build_system=cmake build_type=RelWithDebInfo generator=make patches=0b5dd6f arch=linux-sles15-zen3
[+]  mj6xxft          ^cmake@3.26.3%gcc@11.2.0~doc+ncurses+ownlibs~qt build_system=generic build_type=Release arch=linux-sles15-zen3
[+]  6b72ibn          ^gmake@4.4.1%gcc@11.2.0~guile build_system=autotools arch=linux-sles15-zen3
[+]  savxweu          ^pkgconf@1.8.0%gcc@11.2.0 build_system=autotools arch=linux-sles15-zen3
[+]  vf4m3za      ^hip@5.3.0%gcc@11.2.0~cuda~ipo+rocm build_system=cmake build_type=Release generator=make patches=ca523f1 arch=linux-sles15-zen3
[+]  pgtgquv      ^hipblas@5.3.0%gcc@11.2.0~cuda~ipo+rocm amdgpu_target=auto build_system=cmake build_type=Release generator=make arch=linux-sles15-zen3
[+]  nuk4cfp      ^hipsolver@5.3.0%gcc@11.2.0~cuda~ipo+rocm amdgpu_target=auto build_system=cmake build_type=Release generator=make arch=linux-sles15-zen3
[+]  vi5uhzb      ^hipsparse@5.3.0%gcc@11.2.0~cuda~ipo+rocm amdgpu_target=auto build_system=cmake build_type=Release generator=make patches=c447537 arch=linux-sles15-zen3
[+]  ssqrunh      ^hsa-rocr-dev@5.3.0%gcc@11.2.0+image~ipo+shared build_system=cmake build_type=Release generator=make patches=71e6851 arch=linux-sles15-zen3
[+]  wrk4dxl      ^hypre@2.28.0%gcc@11.2.0~complex~cuda~debug+fortran~gptune~int64~internal-superlu~mixedint+mpi~openmp~rocm+shared~superlu-dist~sycl~umpire~unified-memory build_system=autotools arch=linux-sles15-zen3
[+]  xikyk2u      ^llvm-amdgpu@5.3.0%gcc@11.2.0~ipo~link_llvm_dylib~llvm_dylib~openmp+rocm-device-libs build_system=cmake build_type=Release generator=ninja patches=a08bbe1 arch=linux-sles15-zen3
[+]  smgewc5      ^metis@5.1.0%gcc@11.2.0~gdb~int64~ipo~real64+shared build_system=cmake build_type=RelWithDebInfo generator=make patches=4991da9,93a7903,b1225da arch=linux-sles15-zen3
[+]  qdqswxb      ^openblas@0.3.23%gcc@11.2.0~bignuma~consistent_fpcsr+fortran~ilp64+locking+pic+shared build_system=makefile symbol_suffix=none threads=openmp arch=linux-sles15-zen3
[+]  fr2owxn          ^perl@5.36.0%gcc@11.2.0+cpanm+open+shared+threads build_system=generic arch=linux-sles15-zen3
[+]  gqdvawb              ^berkeley-db@18.1.40%gcc@11.2.0+cxx~docs+stl build_system=autotools patches=26090f4,b231fcc arch=linux-sles15-zen3
[+]  isbuqtz      ^parmetis@4.0.3%gcc@11.2.0~gdb~int64~ipo+shared build_system=cmake build_type=RelWithDebInfo generator=make patches=4f89253,50ed208,704b84f arch=linux-sles15-zen3
[+]  b4zevxj      ^python@3.7.15%gcc@11.2.0+bz2+crypt+ctypes+dbm~debug+libxml2+lzma~nis~optimizations+pic+pyexpat+pythoncmd+readline+shared+sqlite3+ssl~tkinter+uuid+zlib build_system=generic patches=0d98e93,f2fd060 arch=linux-sles15-zen3
[+]  gcwem7x          ^bzip2@1.0.8%gcc@11.2.0~debug~pic+shared build_system=generic arch=linux-sles15-zen3
[+]  uo4mkvg          ^expat@2.5.0%gcc@11.2.0+libbsd build_system=autotools arch=linux-sles15-zen3
[+]  svce26u              ^libbsd@0.11.7%gcc@11.2.0 build_system=autotools arch=linux-sles15-zen3
[+]  bcya2vp                  ^libmd@1.0.4%gcc@11.2.0 build_system=autotools arch=linux-sles15-zen3
[+]  lggkkez          ^gdbm@1.23%gcc@11.2.0 build_system=autotools arch=linux-sles15-zen3
[+]  ug2ciqw          ^gettext@0.21.1%gcc@11.2.0+bzip2+curses+git~libunistring+libxml2+tar+xz build_system=autotools arch=linux-sles15-zen3
[+]  op6kuq4              ^libxml2@2.10.3%gcc@11.2.0~python build_system=autotools arch=linux-sles15-zen3
[+]  uttw4m2              ^tar@1.34%gcc@11.2.0 build_system=autotools zip=pigz arch=linux-sles15-zen3
[+]  eqmk7ji                  ^pigz@2.7%gcc@11.2.0 build_system=makefile arch=linux-sles15-zen3
[+]  yiosjsu                  ^zstd@1.5.5%gcc@11.2.0+programs build_system=makefile compression=none libs=shared,static arch=linux-sles15-zen3
[+]  4zpefmz          ^libffi@3.4.4%gcc@11.2.0 build_system=autotools arch=linux-sles15-zen3
[+]  xaieuhm          ^libxcrypt@4.4.33%gcc@11.2.0~obsolete_api build_system=autotools arch=linux-sles15-zen3
[+]  igbrz2c          ^ncurses@6.3%gcc@11.2.0~symlinks+termlib abi=none build_system=autotools arch=linux-sles15-zen3
[+]  hr2fmma          ^openssl@1.1.1-os%gcc@11.2.0~docs~shared build_system=generic certs=mozilla arch=linux-sles15-zen3
[+]  rmgt2pq          ^readline@8.2%gcc@11.2.0 build_system=autotools patches=bbf97f1 arch=linux-sles15-zen3
[+]  eaxessg          ^sqlite@3.40.1%gcc@11.2.0+column_metadata+dynamic_extensions+fts~functions+rtree build_system=autotools arch=linux-sles15-zen3
[+]  4qyr4mp          ^util-linux-uuid@2.38.1%gcc@11.2.0 build_system=autotools arch=linux-sles15-zen3
[+]  yxqlxfz          ^xz@5.4.1%gcc@11.2.0+pic build_system=autotools libs=shared,static arch=linux-sles15-zen3
[+]  hi3lc3e      ^rocblas@5.3.0%gcc@11.2.0~ipo+tensile amdgpu_target=auto build_system=cmake build_type=Release generator=make patches=81591d9 arch=linux-sles15-zen3
[+]  5uu3uc7      ^rocprim@5.3.0%gcc@11.2.0~ipo amdgpu_target=auto build_system=cmake build_type=Release generator=make patches=98506fb arch=linux-sles15-zen3
[+]  a3ylrjj      ^rocrand@5.3.0%gcc@11.2.0+hiprand~ipo amdgpu_target=auto build_system=cmake build_type=Release generator=make patches=a35e689 arch=linux-sles15-zen3
[+]  oksm2jz      ^rocsolver@5.3.0%gcc@11.2.0~ipo+optimal amdgpu_target=auto build_system=cmake build_type=Release generator=make patches=8067bfb arch=linux-sles15-zen3
[+]  mhzqnrk      ^rocsparse@5.3.0%gcc@11.2.0~ipo~test amdgpu_target=auto build_system=cmake build_type=Release generator=make arch=linux-sles15-zen3
[+]  uxs2dfz      ^rocthrust@5.3.0%gcc@11.2.0~ipo amdgpu_target=auto build_system=cmake build_type=Release generator=make arch=linux-sles15-zen3
[+]  v22twoq      ^superlu-dist@8.1.2%gcc@11.2.0~cuda~int64~ipo~openmp~rocm+shared build_system=cmake build_type=RelWithDebInfo generator=make arch=linux-sles15-zen3
[+]  d35p7mx      ^zlib@1.2.13%gcc@11.2.0+optimize+pic+shared build_system=makefile arch=linux-sles15-zen3
</pre></details>



Error:
```
$> spack -e . install
...
==> Installing petsc-3.19.1-3xruui4hl7p4jk7mkm23qeswtty4lkul
==> No binary for petsc-3.19.1-3xruui4hl7p4jk7mkm23qeswtty4lkul found: installing from source
==> Using cached archive: /tmp/lpeyrala/PrgEnv-gnu/spack/var/spack/cache/_source-cache/archive/74/74db60c53c80b48d5c39e07bc39a883ecced88b9f24a5de17cf6f485a903e120.tar.gz
==> No patches needed for petsc
==> petsc: Executing phase: 'configure'
==> petsc: Executing phase: 'build'
==> Error: ProcessError: Command exited with status 2:
    'make' '-j8' 'V=1'

5 errors found in build log:
     1767    In file included from /opt/rocm-5.3.0/include/thrust/scan.h:1656:
     1768    In file included from /opt/rocm-5.3.0/include/thrust/detail/scan.inl:28:
     1769    In file included from /opt/rocm-5.3.0/include/thrust/system/detail/adl/scan.h:44:
     1770    In file included from /opt/rocm-5.3.0/include/thrust/system/hip/detail/scan.h:48:
     1771    In file included from /opt/rocm-5.3.0/include/rocprim/rocprim.hpp:55:
     1772    In file included from /opt/rocm-5.3.0/include/rocprim/device/device_merge.hpp:31:
  >> 1773    /opt/rocm-5.3.0/include/rocprim/device/detail/device_merge.hpp:133:9: error: no matching function for call to 'merge_path'
     1774            merge_path(
     1775            ^~~~~~~~~~
     1776    /opt/rocm-5.3.0/include/rocprim/device/device_merge.hpp:57:5: note: in instantiation of function template specialization 'rocprim::detail::partition
             _kernel_impl<unsigned int *, thrust::zip_iterator<thrust::tuple<thrust::detail::normal_iterator<thrust::device_ptr<int>>, thrust::detail::normal_ite
             rator<thrust::device_ptr<int>>, thrust::detail::normal_iterator<thrust::device_ptr<double>>, thrust::constant_iterator<int>>>, thrust::zip_iterator<
             thrust::tuple<thrust::detail::normal_iterator<thrust::device_ptr<int>>, thrust::transform_iterator<Shift, thrust::detail::normal_iterator<thrust::de
             vice_ptr<int>>>, thrust::detail::normal_iterator<thrust::device_ptr<double>>, thrust::constant_iterator<int>>>, IJCompare4>' requested here
     1777        partition_kernel_impl(
     1778        ^
     1779    /opt/rocm-5.3.0/include/rocprim/device/device_merge.hpp:180:33: note: in instantiation of function template specialization 'rocprim::detail::partiti
             on_kernel<unsigned int *, thrust::zip_iterator<thrust::tuple<thrust::detail::normal_iterator<thrust::device_ptr<int>>, thrust::detail::normal_iterat
             or<thrust::device_ptr<int>>, thrust::detail::normal_iterator<thrust::device_ptr<double>>, thrust::constant_iterator<int>>>, thrust::zip_iterator<thr
             ust::tuple<thrust::detail::normal_iterator<thrust::device_ptr<int>>, thrust::transform_iterator<Shift, thrust::detail::normal_iterator<thrust::devic
             e_ptr<int>>>, thrust::detail::normal_iterator<thrust::device_ptr<double>>, thrust::constant_iterator<int>>>, IJCompare4>' requested here

     ...

     1798          PetscCallThrust(thrust::merge(thrust::device, Azb, Aze, Bzb, Bze, Czb, IJCompare4()));
     1799                                  ^
     1800    /opt/rocm-5.3.0/include/rocprim/device/detail/device_merge.hpp:77:9: note: candidate template ignored: deduced conflicting types for parameter 'Keys
             InputIterator' ('zip_iterator<tuple<[...], thrust::detail::normal_iterator<thrust::device_ptr<int>>, [8 * ...]>>' vs. 'zip_iterator<tuple<[...], thr
             ust::transform_iterator<Shift, thrust::detail::normal_iterator<thrust::device_ptr<int>>>, [8 * ...]>>')
     1801    OffsetT merge_path(KeysInputIterator keys_input1,
     1802            ^
     1803    1 error generated when compiling for gfx90a.
  >> 1804    make[3]: *** [gmakefile:204: arch-linux-c-opt/obj/mat/impls/aij/seq/seqhipsparse/aijhipsparse.o] Error 1
     1805    make[3]: *** Waiting for unfinished jobs....
...
```

### Error message

<details><summary>Error message</summary>

<pre>
... see above
</pre></details>


### Information on your system

* **Spack:** 0.20.0.dev0 (6ebc3f0b011a90752b602d4d3a2e215ff15c4a2c)
* **Python:** 3.6.15
* **Platform:** linux-sles15-zen3
* **Concretizer:** clingo

### Additional information

[configure.log](https://github.com/spack/spack/files/11391024/configure.log)
[spack-build-env.txt](https://github.com/spack/spack/files/11391027/spack-build-env.txt)
[spack-build-out.txt](https://github.com/spack/spack/files/11391028/spack-build-out.txt)

@balay @barrysmith @jedbrown @wspear

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have run `spack maintainers <name-of-the-package>` and **@mentioned** any maintainers
- [X] I have uploaded the build log and environment files
- [X] I have searched the issues of this repo and believe this is not a duplicate

--- a/var/spack/repos/builtin/packages/petsc/package.py
+++ b/var/spack/repos/builtin/packages/petsc/package.py
@@ -145,7 +145,7 @@ class Petsc(Package, CudaPackage, ROCmPackage):
     variant("fortran", default=True, description="Activates fortran support")
 
     # https://github.com/spack/spack/issues/37416
-    conflicts("^rocprim@5.3.0", when="@3.19.1 +rocm")
+    conflicts("^rocprim@5.3.0:5.3.2", when="+rocm")
 
     # 3.8.0 has a build issue with MKL - so list this conflict explicitly
     conflicts("^intel-mkl", when="@3.8.0")

--- a/var/spack/repos/builtin/packages/petsc/package.py
+++ b/var/spack/repos/builtin/packages/petsc/package.py
@@ -144,6 +144,9 @@ class Petsc(Package, CudaPackage, ROCmPackage):
     variant("kokkos", default=False, description="Activates support for kokkos and kokkos-kernels")
     variant("fortran", default=True, description="Activates fortran support")
 
+    # https://github.com/spack/spack/issues/37416
+    conflicts("^rocprim@5.3.0", when="@3.19.1 +rocm")
+
     # 3.8.0 has a build issue with MKL - so list this conflict explicitly
     conflicts("^intel-mkl", when="@3.8.0")
 

